quic/s2n-quic-core/src/inet/ipv6.rs

```diff
@@ -377,6 +386,196 @@ impl From<IpV6Address> for [u16; IPV6_LEN / 2] {
+//#   |                                                               |
+//#   +                                                               +
+//#   |                                                               |
+//#   +                      Destination Address                      +
+//#   |                                                               |
+//#   +                                                               +
+//#   |                                                               |
+//#   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+define_inet_type!(
+    pub struct Header {
+        vtf: Vtf,
+        payload_len: U16,
+        next_header: ip::Protocol,
+        hop_limit: u8,
+        source: IpV6Address,
+        destination: IpV6Address,
+    }
+);
+
+impl fmt::Debug for Header {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("ipv6::Header")
+            .field("version", &self.vtf.version())
+            .field("dscp", &self.vtf.dscp())
+            .field("ecn", &self.vtf.ecn())
+            .field(
+                "flow_label",
+                &format_args!("0x{:05x}", self.vtf.flow_label()),
+            )
+            .field("payload_len", &self.payload_len)
+            .field("next_header", &self.next_header)
+            .field("hop_limit", &self.hop_limit)
+            .field("source", &self.source)
+            .field("destination", &self.destination)
+            .finish()
+    }
+}
+
+impl Header {
+    /// Swaps the direction of the header
+    #[inline]
+    pub fn swap(&mut self) {
+        core::mem::swap(&mut self.source, &mut self.destination)
+    }
+
+    #[inline]
+    pub const fn vtf(&self) -> &Vtf {
+        &self.vtf
+    }
+
+    #[inline]
+    pub fn vtf_mut(&mut self) -> &mut Vtf {
+        &mut self.vtf
+    }
+
+    #[inline]
+    pub const fn payload_len(&self) -> &U16 {
+        &self.payload_len
+    }
+
+    #[inline]
+    pub fn payload_len_mut(&mut self) -> &mut U16 {
+        &mut self.payload_len
+    }
+
+    #[inline]
+    pub const fn next_header(&self) -> &ip::Protocol {
+        &self.next_header
+    }
+
+    #[inline]
+    pub fn next_header_mut(&mut self) -> &mut ip::Protocol {
+        &mut self.next_header
+    }
+
+    #[inline]
+    pub const fn hop_limit(&self) -> &u8 {
+        &self.hop_limit
+    }
+
+    #[inline]
+    pub fn hop_limit_mut(&mut self) -> &mut u8 {
+        &mut self.hop_limit
+    }
+
+    #[inline]
+    pub const fn source(&self) -> &IpV6Address {
+        &self.source
+    }
+
+    #[inline]
+    pub fn source_mut(&mut self) -> &mut IpV6Address {
+        &mut self.source
+    }
+
+    #[inline]
+    pub const fn destination(&self) -> &IpV6Address {
+        &self.destination
+    }
+
+    #[inline]
+    pub fn destination_mut(&mut self) -> &mut IpV6Address {
+        &mut self.destination
+    }
+}
+
+define_inet_type!(
```

Comment:
Same thing about opaque boundaries (calling out in a comment that the version field is the 4 most significant bits, etc), especially since the RFC diagram doesn't include the field lengths.

Reply:
Actually, nevermind, it apparently does specify lengths. I just counted that flow label was 20 bits long and assumed that wasn't accurate. So just explicitly calling that out would be nice.